### PR TITLE
[IOTDB-5528] Refactor schema engine and region statistics

### PR DESCRIPTION
**Pull request by Cpaulyz**
## Description

The memory statistics for the current metadata are global. For better monitoring, the granularity of memory statistics needs to be refined per SchemaRegion.

image

See https://apache-iotdb.feishu.cn/docx/HD2hdIYS6oBSzsxpFw2cCNPFnoe for detail

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/metadata/rescon/SchemaResourceManager.java [hunk 1]
```diff
@@ -44,8 +43,6 @@ public static void initSchemaResource() {
   }
 
   public static void clearSchemaResource() {
-    SchemaStatisticsManager.getInstance().clear();
-    MemoryStatistics.getInstance().clear();
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/metadata/rescon/SchemaResourceManager.java:48]: SchemaEngineStatistics will be clear after clear all schema region

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionSchemaFileImpl.java [hunk 2]
```diff
@@ -413,8 +418,7 @@ public SchemaRegionId getSchemaRegionId() {
   @Override
   public synchronized void deleteSchemaRegion() throws MetadataException {
     // collect all the LeafMNode in this schema region
-    long seriesCount = mtree.countAllMeasurement();
-    schemaStatisticsManager.deleteTimeseries(seriesCount);
+    long seriesCount = regionStatistics.getSeriesNumber();
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionSchemaFileImpl.java:421]: release resources in `clear()`

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionMemoryImpl.java [hunk 3]
```diff
@@ -378,8 +382,7 @@ public SchemaRegionId getSchemaRegionId() {
   @Override
   public synchronized void deleteSchemaRegion() throws MetadataException {
     // collect all the LeafMNode in this schema region
-    long seriesCount = mtree.countAllMeasurement();
-    schemaStatisticsManager.deleteTimeseries(seriesCount);
+    long seriesCount = regionStatistics.getSeriesNumber();
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionMemoryImpl.java:385]: release resources in `clear()`

File: server/src/main/java/org/apache/iotdb/db/metadata/rescon/MemSchemaEngineStatistics.java [hunk 4]
```diff
@@ -26,72 +25,84 @@
...
   public long getMemoryUsage() {
     return memoryUsage.get();
   }
 
   public void requestMemory(long size) {
-    memoryUsage.getAndUpdate(v -> v += size);
+    memoryUsage.addAndGet(size);
     if (memoryUsage.get() >= memoryCapacity) {
-      logger.warn("Current series number {} is too large...", memoryUsage);
+      logger.warn("Current series memory {} is too large...", memoryUsage);
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/metadata/rescon/MemSchemaEngineStatistics.java:73]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/rescon/CachedSchemaEngineStatistics.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ * This class is used to record global statistics for SchemaEngine in Schema_File mode, which is a
+ * superset of the statistics in Memory mode
+ */
+public class CachedSchemaEngineStatistics extends MemSchemaEngineStatistics {
+
+  private final AtomicLong unpinnedSize = new AtomicLong(0);
+  private final AtomicLong pinnedSize = new AtomicLong(0);
+  private final AtomicLong unpinnedNum = new AtomicLong(0);
+  private final AtomicLong pinnedNum = new AtomicLong(0);
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/metadata/rescon/CachedSchemaEngineStatistics.java:37]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/rescon/CachedSchemaRegionStatistics.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * This class is used to record statistics within a SchemaRegion in Schema_File mode, which is a
+ * superset of the statistics in Memory mode
+ */
+public class CachedSchemaRegionStatistics extends MemSchemaRegionStatistics {
+
+  private final AtomicLong unpinnedSize = new AtomicLong(0);
+  private final AtomicLong pinnedSize = new AtomicLong(0);
+  private final AtomicLong unpinnedNum = new AtomicLong(0);
+  private final AtomicLong pinnedNum = new AtomicLong(0);
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/metadata/rescon/CachedSchemaRegionStatistics.java:32]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/CachedMTreeStore.java [hunk 7]
```diff
@@ -483,7 +522,7 @@ public StampedWriterPreferredLock getLock() {
    * no node could be evicted. Update the memory status after evicting each node.
    */
   public void executeMemoryRelease() {
-    while (memManager.isExceedReleaseThreshold() && !memManager.isEmpty()) {
+    while (regionStatistics.isExceedReleaseThreshold() && regionStatistics.getCachedSize() != 0) {
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/CachedMTreeStore.java:525]: move releaseFlushStrategy to CacheMemoryManager

[the remark this answers is not in the document]
Comment (MarcosZyk) [reply server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/CachedMTreeStore.java]: Should this judgement be transferred to CacheMemoryManager too? It seems the store only needs to provide how to release node but not when to end release?

**Review by MarcosZyk — APPROVED** [review]

**Pull request merged**